src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,420 @@
+import com.minecolonies.coremod.colony.jobs.AbstractJob;
+import com.minecolonies.coremod.entity.ai.basic.AbstractEntityAIInteract;
+import com.minecolonies.coremod.entity.ai.util.AIState;
+import com.minecolonies.coremod.entity.ai.util.AITarget;
+import com.minecolonies.coremod.util.StructureWrapper;
+import net.minecraft.entity.passive.EntityAnimal;
+import net.minecraft.init.Items;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.DamageSource;
+import net.minecraft.util.Mirror;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.TextComponentTranslation;
+import net.minecraftforge.items.wrapper.InvWrapper;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.minecolonies.api.util.constant.ToolLevelConstants.TOOL_LEVEL_WOOD_OR_GOLD;
+import static com.minecolonies.coremod.entity.ai.util.AIState.*;
+
+/**
+ * Abstract class for all Citizen Herder AIs
+ */
+public abstract class AbstractEntityAIHerder<J extends AbstractJob, T extends EntityAnimal> extends AbstractEntityAIInteract<J>
+{
+
+    /**
+     * Amount of animals needed to bread.
+     */
+    public static final int NUM_OF_ANIMALS_TO_BREED = 2;
+
+    /**
+     * Butchering attack damage.
+     */
+    private static final int BUTCHERING_ATTACK_DAMAGE = 5;
+
+    /**
+     * Distance two animals need to be inside to breed.
+     */
+    private static final int DISTANCE_TO_BREED = 10;
+
+    /**
+     * Amount of animals to keep per Hut Level.
+     */
+    private static int maxAnimalMultiplier;
+
+    /**
+     * Creates the abstract part of the AI.
+     * Always use this constructor!
+     *
+     * @param job the job to fulfill
+     */
+    public AbstractEntityAIHerder(@NotNull final J job, final int maxAnimalsMultiplier)
+    {
+        super(job);
+        maxAnimalMultiplier = maxAnimalsMultiplier;
+        super.registerTargets(
+          new AITarget(IDLE, START_WORKING),
+          new AITarget(START_WORKING, this::startWorkingAtOwnBuilding),
+          new AITarget(PREPARING, this::prepareForHerding),
+          new AITarget(HERDER_DECIDE, this::decideWhatToDo),
+          new AITarget(HERDER_BREED, this::breedAnimals),
+          new AITarget(HERDER_BUTCHER, this::butcherAnimals)
+        );
+        worker.setCanPickUpLoot(true);
+    }
+
+    @Override
+    protected int getActionsDoneUntilDumping()
+    {
+        return 10;
+    }
+
+    /**
+     * Decides what job the herder should switch to, breeding, butchering, or shearing (If Shepherd).
+     *
+     * @return the next AIState the shepherd should switch to, after executing this method.
+     */
+    public AIState decideWhatToDo()
+    {
+        setDelay(40);
+
+        final List<T> animals = new ArrayList<>(getAnimals());
+
+        if (animals.isEmpty())
+        {
+            setDelay(100);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 100 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)